### PR TITLE
Refactor to use AddingTrailingDataSubscriber

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Refactor `CompressionAsyncRequestBody` to use [`AddingTrailingDataSubscriber`](https://github.com/aws/aws-sdk-java-v2/pull/4366/) and remove duplicate logic

## Modifications

Replace existing logic to send buffered data with `addTrailingData()`

## Testing

All current tests pass

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java [hunk 1]
```diff
@@ -140,73 +150,16 @@ public Builder chunkSize(Integer chunkSize) {
 
     private static final class SplittingSubscriber extends DelegatingSubscriber<ByteBuffer, Iterable<ByteBuffer>> {
         private final ChunkBuffer chunkBuffer;
-        private final AtomicBoolean upstreamDone = new AtomicBoolean(false);
-        private final AtomicLong downstreamDemand = new AtomicLong();
-        private final Object lock = new Object();
-        private volatile boolean sentFinalChunk = false;
 
-        protected SplittingSubscriber(Subscriber<? super Iterable<ByteBuffer>> subscriber, int chunkSize) {
+        protected SplittingSubscriber(Subscriber<? super Iterable<ByteBuffer>> subscriber, ChunkBuffer chunkBuffer) {
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java:154]: Do we need `chunkBuffer` in this class?

Comment (davidh44) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java]: Right not needed, removed from SplittingSubscriber

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java [hunk 2]
```diff
@@ -70,7 +71,16 @@ public String contentType() {
     }
 
     private SdkPublisher<Iterable<ByteBuffer>> split(SdkPublisher<ByteBuffer> source) {
-        return subscriber -> source.subscribe(new SplittingSubscriber(subscriber, chunkSize));
+        return subscriber -> source.subscribe(new SplittingSubscriber(subscriber));
+    }
+
+    private Iterable<ByteBuffer> getBufferedDataIfPresent() {
```
Comment (joviegas) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java:77]: Can we use this inline ?
```java
private Iterable<ByteBuffer> getBufferedDataIfPresent() {
    return chunkBuffer.getBufferedData()
            .map(Collections::singletonList)
            .orElse(Collections.emptyList());
}
```

Comment (davidh44) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/CompressionAsyncRequestBody.java:77]: Good suggestion, updated

**Review by joviegas — APPROVED** [review]

**Pull request merged**